components/usage/pkg/db/cost_center.go

```diff
@@ -102,57 +105,112 @@ func getCostCenter(ctx context.Context, conn *gorm.DB, attributionId Attribution
 	return costCenter, nil
 }
 
-func (c *CostCenterManager) UpdateCostCenter(ctx context.Context, costCenter CostCenter) (CostCenter, error) {
+func (c *CostCenterManager) UpdateCostCenter(ctx context.Context, newCC CostCenter) (CostCenter, error) {
+	if newCC.SpendingLimit < 0 {
+		return CostCenter{}, status.Errorf(codes.InvalidArgument, "Spending limit cannot be set below zero.")
+	}
 
+	attributionID := newCC.ID
 	// retrieving the existing cost center to maintain the readonly values
-	existingCostCenter, err := c.GetOrCreateCostCenter(ctx, costCenter.ID)
+	existingCC, err := c.GetOrCreateCostCenter(ctx, newCC.ID)
 	if err != nil {
-		return CostCenter{}, err
+		return CostCenter{}, status.Errorf(codes.NotFound, "cost center does not exist")
 	}
 
 	now := time.Now()
 
 	// we always update the creationTime
-	costCenter.CreationTime = NewVarcharTime(now)
+	newCC.CreationTime = NewVarcharTime(now)
 	// we don't allow setting the nextBillingTime from outside
-	costCenter.NextBillingTime = existingCostCenter.NextBillingTime
-
-	// Do we have a billing strategy update?
-	if costCenter.BillingStrategy != existingCostCenter.BillingStrategy {
-		switch costCenter.BillingStrategy {
-		case CostCenter_Stripe:
-			// moving to stripe -> let's run a finalization
-			finalizationUsage, err := c.ComputeInvoiceUsageRecord(ctx, costCenter.ID)
+	newCC.NextBillingTime = existingCC.NextBillingTime
+
+	isTeam := attributionID.IsEntity(AttributionEntity_Team)
+	isUser := attributionID.IsEntity(AttributionEntity_User)
+
+	if isUser {
+		// New billing strategy is Stripe
+		if newCC.BillingStrategy == CostCenter_Stripe {
+			if newCC.SpendingLimit < c.cfg.MinForIndividualsOnStripe {
+				return CostCenter{}, status.Errorf(codes.FailedPrecondition, "individual users cannot lower their spending below %d", c.cfg.ForUsers)
+			}
+		}
+
+		// Billing strategy remains unchanged (Other)
+		if newCC.BillingStrategy == CostCenter_Other && existingCC.BillingStrategy == CostCenter_Other {
+			if newCC.SpendingLimit != existingCC.SpendingLimit {
+				return CostCenter{}, status.Errorf(codes.FailedPrecondition, "individual users on a free plan cannot adjust their spending limit")
+			}
+		}
+
+		// Downgrading from stripe
+		if existingCC.BillingStrategy == CostCenter_Stripe && newCC.BillingStrategy == CostCenter_Other {
+			newCC.SpendingLimit = c.cfg.ForUsers
```

Comment:
Hmm, this flow still doesn't seem to work properly 🤔

1. Create a team called "Gitpod [Something]"
2. Go to `/billing` (you have 500 credits 👍)

<details><summary>(view DB state)</summary>

```
mysql> select * from d_b_cost_center where id = 'user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7';
+-------------------------------------------+---------------+---------+----------------------------+--------------------------+-----------------+--------------------------+
| id                                        | spendingLimit | deleted | _lastModified              | creationTime             | billingStrategy | nextBillingTime          |
+-------------------------------------------+---------------+---------+----------------------------+--------------------------+-----------------+--------------------------+
| user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7 |           500 |       0 | 2022-10-13 08:59:20.491162 | 2022-10-13T08:59:20.488Z | other           | 2022-11-13T08:59:20.489Z |
+-------------------------------------------+---------------+---------+----------------------------+--------------------------+-----------------+--------------------------+
1 row in set (0.01 sec)
```

</details>

3. Upgrade by adding a payment method, then refresh (limit is 100 ? 😳 Should be 1000)

<details><summary>(view DB state)</summary>

```
mysql> select * from d_b_cost_center where id = 'user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7';
+-------------------------------------------+---------------+---------+----------------------------+--------------------------+-----------------+--------------------------+
| id                                        | spendingLimit | deleted | _lastModified              | creationTime             | billingStrategy | nextBillingTime          |
+-------------------------------------------+---------------+---------+----------------------------+--------------------------+-----------------+--------------------------+
| user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7 |           500 |       0 | 2022-10-13 08:59:20.491162 | 2022-10-13T08:59:20.488Z | other           | 2022-11-13T08:59:20.489Z |
| user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7 |           100 |       0 | 2022-10-13 09:01:33.580476 | 2022-10-13T09:01:33.576Z | stripe          |                          |
+-------------------------------------------+---------------+---------+----------------------------+--------------------------+-----------------+--------------------------+
2 rows in set (0.02 sec)
```

<details>

4. Try to set the limit to 150 -- it should fail because that's < 1000, but actually it works 😳

<details><summary>(view DB state)</summary>

```
mysql> select * from d_b_cost_center where id = 'user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7';
+-------------------------------------------+---------------+---------+----------------------------+--------------------------+-----------------+--------------------------+
| id                                        | spendingLimit | deleted | _lastModified              | creationTime             | billingStrategy | nextBillingTime          |
+-------------------------------------------+---------------+---------+----------------------------+--------------------------+-----------------+--------------------------+
| user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7 |           500 |       0 | 2022-10-13 08:59:20.491162 | 2022-10-13T08:59:20.488Z | other           | 2022-11-13T08:59:20.489Z |
| user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7 |           100 |       0 | 2022-10-13 09:01:33.580476 | 2022-10-13T09:01:33.576Z | stripe          |                          |
| user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7 |           100 |       0 | 2022-10-13 09:03:49.189822 | 2022-10-13T09:03:49.186Z | stripe          |                          |
| user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7 |           150 |       0 | 2022-10-13 09:03:53.190384 | 2022-10-13T09:03:53.187Z | stripe          |                          |
+-------------------------------------------+---------------+---------+----------------------------+--------------------------+-----------------+--------------------------+
4 rows in set (0.02 sec)
```

</details>

5. Manage plan, then cancel your Stripe subscription -- the limit should reset to 500, but instead it remains at its previous value 🐛

<details><summary>(view DB state)</summary>

```
mysql> select * from d_b_cost_center where id = 'user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7';
+-------------------------------------------+---------------+---------+----------------------------+--------------------------+-----------------+--------------------------+
| id                                        | spendingLimit | deleted | _lastModified              | creationTime             | billingStrategy | nextBillingTime          |
+-------------------------------------------+---------------+---------+----------------------------+--------------------------+-----------------+--------------------------+
| user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7 |           500 |       0 | 2022-10-13 08:59:20.491162 | 2022-10-13T08:59:20.488Z | other           | 2022-11-13T08:59:20.489Z |
| user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7 |           100 |       0 | 2022-10-13 09:01:33.580476 | 2022-10-13T09:01:33.576Z | stripe          |                          |
| user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7 |           100 |       0 | 2022-10-13 09:03:49.189822 | 2022-10-13T09:03:49.186Z | stripe          |                          |
| user:f0f6f6f5-f8e4-47e4-80f9-420f8cbcbfb7 |           150 |       0 | 2022-10-13 09:03:53.190384 | 2022-10-13T09:03:53.187Z | stripe          |                          |
+-------------------------------------------+---------------+---------+----------------------------+--------------------------+-----------------+--------------------------+
4 rows in set (0.01 sec)
```

Reply:
Aha, so it's even 0 in preview environments 👌 thanks for clarifying.

Still, this doesn't explain why, upon cancellation, the user cost center does not go back to `other` and doesn't reset its usage limit to `500`.

Reply:
But shouldn't that happen automatically when Stripe calls Gitpod to tell it that a subscription was cancelled?

Reply:
Setting the config (now done in preview also) addresses 3. not failing. 

The default 100 comes from `protected defaultSpendingLimit = 100;` defined in gitpod-server-impl.ts